[HttpKernel] Add `ControllerEvent::getAttributes()` to handle attributes on controllers

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Replacing #45457. Paving the way toward #44705

nicolas-grekas committed:
[HttpKernel] Add `ControllerEvent::getAttributes()` to handle attributes on controllers

lyrixx commented:
I wanted do that for ages! Thanks for doing it. But iirc we already had an instance of the reflection somewhere. Can't we reuse it?

GromNaN commented on src/Symfony/Component/HttpKernel/Event/ControllerEvent.php:
```diff
@@ -45,4 +46,31 @@ public function setController(callable $controller): void
+        return $this->attributes;
+    }
+
+    public function setAttributes(array $attributes): void
```
What is the use-case for the setter ?

nicolas-grekas replied:
To allow injecting attributes (value objects) that are defined in any other way than native php attributes.

lyrixx replied:
I'm not sure it's a good idea to introduce that for now

nicolas-grekas replied:
I think it's important to give this flexibility. Native attributes must remain one way among any other, and this is the only way. open/closed principle.

nicolas-grekas commented:
> we already had an instance of the reflection somewhere. Can't we reuse it?

yes we can, updated!

stof commented on src/Symfony/Component/HttpKernel/Event/ControllerEvent.php:
```diff
@@ -43,6 +44,30 @@ public function getController(): callable
+            }
+        }
+
+        return $this->attributes;
```
to me, this is missing an assignment before the loop, to ensure that it is an array even if the loop does not do any assignment

stof commented on src/Symfony/Component/HttpKernel/Event/ControllerEvent.php:
```diff
@@ -43,6 +44,32 @@ public function getController(): callable
+        return $this->attributes;
+    }
+
+    public function setAttributes(array $attributes): void
```
This should describe the expected structure of the array

stof commented on src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php:
```diff
@@ -45,8 +45,9 @@ public function __construct(ArgumentMetadataFactoryInterface $argumentMetadataFa
     public function getArguments(Request $request, callable $controller): array
     {
         $arguments = [];
+        $reflectors = $request->attributes->get('_controller_reflectors') ?? [];
```
if `$event->setAttributes` is used, I don't see where they would be assigned there

nicolas-grekas replied:
They would always be assigned, because ControllerEvent::setController is always called, and this is independent from using setAttributes or not. Or do you mean something else?

stof replied:
but if `setController` is called before `setAttributes` (likely to be the case if a listener is the one calling `setAttributes` to set custom attributes), the assignment done in `setController` will not include those custom attributes.
To me, the kernel should be the one doing that assignment *after* triggering the event, based on `$event->getAttributes`

stof commented on src/Symfony/Component/HttpKernel/HttpKernel.php:
```diff
@@ -142,6 +142,7 @@ private function handleRaw(Request $request, int $type = self::MAIN_REQUEST): Re
         $arguments = $this->argumentResolver->getArguments($request, $controller);
+        $request->attributes->remove('_controller_reflectors');
 
         $event = new ControllerArgumentsEvent($this, $controller, $arguments, $request, $type);
```
I think the attributes being added in the ControllerEvent need to be exposed (as read-only) in the ControllerArgumentsEvent. Otherwise, those attributes are available for argument resolvers but not for listeners (like the listener handling `IsGranted`)

nicolas-grekas replied:
I'm not sure I see what you mean: to me, the listener handling IsGranted will listen on ControllerEvent and call getAttributes on it. Why would ControllerArgumentsEvent need access to the attributes of the controller itself (compared to only the arguments')? To me, such a listener should listen to ControllerEvent.

stof replied:
the IsGrantedListener runs on ControllerArgumentsEvent, not on ControllerEvent. The goal is to run *after* argument resolvers, so that you can vote on the resolved arguments.

stof commented on src/Symfony/Component/HttpKernel/Event/ControllerEvent.php:
```diff
@@ -43,6 +44,32 @@ public function getController(): callable
+        return $this->attributes ??= [];
+    }
+
+    public function setAttributes(array $attributes): void
```
should we have a `addAttribute` method instead, that allows adding additional attributes without loosing the ones coming from Reflection ?

nicolas-grekas replied:
This works already if one wants to add an attribute without overriding the existing ones:
`$event->setAttributes(array_merge_recursive('my_attr' => [new Foo()], $event->getAttributes())`

So I would reformulate your question as: is this common enough to deserve a simpler API? I'm not sure personally. WDYT?

stof replied:
well, the need to perform a recursive merge is easy to overlook. And I don't see any use case where you would want to *remove* the attributes coming from Reflection. This would be a big WTF for someone having to debug why their attributes have no effect.

nicolas-grekas commented:
PR is ready for another round of reviews.

@stof, I  addressed your comment this way:
> the attributes being added in the ControllerEvent need to be exposed (as read-only) in the ControllerArgumentsEvent.

> the IsGrantedListener runs on ControllerArgumentsEvent, not on ControllerEvent. The goal is to run after argument resolvers, so that you can vote on the resolved arguments.

Understood and done. Just one thing: because there is `ControllerArgumentsEvent::setController()`, I think we should allow also updating the attributes. This is what the patch does.

> should we have a addAttribute method instead, that allows adding additional attributes without loosing the ones coming from Reflection ?

> well, the need to perform a recursive merge is easy to overlook. And I don't see any use case where you would want to remove the attributes coming from Reflection. This would be a big WTF for someone having to debug why their attributes have no effect.

I replaced `setAttributes()` by a new `$attributes` argument on `setController()` ~and a new `addAttribute()` method~. Note that I think it's important to allow ppl to opt-out from reflection. This is the same principle that we're following everywhere we read attributes: they should remain declarative, which means optional.

nicolas-grekas commented:
I missed this comment:

> if setController is called before setAttributes (likely to be the case if a listener is the one calling setAttributes to set custom attributes), the assignment done in setController will not include those custom attributes.

Having `$attributes` on `setController()` binds controllers and their attributes. I think this solve this concern. But having a separate `addAttribute()` method reintroduces this concern. That's why I've removed it. I think we need to provide a way to override real attributes, but this way should not be a foot-gun. Since it's for advanced use cases, it's fine to me if it's a bit involving.

> To me, the kernel should be the one doing that assignment after triggering the event, based on $event->getAttributes

I want everything to remain lazy, that's what the current implementation provides.

lyrixx commented on src/Symfony/Component/HttpKernel/ControllerMetadata/ArgumentMetadataFactory.php:
```diff
@@ -21,22 +21,15 @@ final class ArgumentMetadataFactory implements ArgumentMetadataFactoryInterface
-            }
+        if (null === $reflection) {
+            $reflection = new \ReflectionFunction($controller(...));
+            $class = str_contains($reflection->name, '{closure}') ? null : $reflection->getClosureScopeClass();
```
At some point, one should write a RFC to add support for `$refection->isClosure()` directly in PHP :) 👼🏼

nicolas-grekas replied:
🔮 https://github.com/php/php-src/pull/8499

lyrixx commented on src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php:
```diff
@@ -28,25 +28,32 @@
     private array $arguments;
 
-    public function __construct(HttpKernelInterface $kernel, callable $controller, array $arguments, Request $request, ?int $requestType)
+    public function __construct(HttpKernelInterface $kernel, callable|ControllerEvent $controller, array $arguments, Request $request, ?int $requestType)
```
Should we deprecate passing a callable as `$controller`?
It'll simplify the code here.

nicolas-grekas replied:
We could deprecate all arguments except $arguments and $controllerEvent indeed. I'll let this as an exercise for someone else if you don't mind ☺️

lyrixx commented on src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php:
```diff
@@ -28,25 +28,32 @@
 final class ControllerArgumentsEvent extends KernelEvent
 {
-    private $controller;
+    private ControllerEvent $controller;
```
I would rename the var to `controllerEvent`.

nicolas-grekas replied:
updated

chalasr commented on src/Symfony/Component/HttpKernel/ControllerMetadata/ArgumentMetadataFactory.php:
```diff
@@ -21,22 +21,15 @@ final class ArgumentMetadataFactory implements ArgumentMetadataFactoryInterface
      * {@inheritdoc}
      */
-    public function createArgumentMetadata(string|object|array $controller): array
+    public function createArgumentMetadata(string|object|array $controller, \ReflectionClass $class = null, \ReflectionFunction $reflection = null): array
```
final + interface ftw 💚

chalasr reviewed (approved):
👍

chalasr commented:
Thank you @nicolas-grekas.

catch56 commented on src/Symfony/Component/HttpKernel/ControllerMetadata/ArgumentMetadataFactory.php:
```diff
@@ -21,22 +21,15 @@ final class ArgumentMetadataFactory implements ArgumentMetadataFactoryInterface
     {
         $arguments = [];
 
-        if (\is_array($controller)) {
```
This has caused a regression on Drupal 10 with the latest Symfony 6.2. Adding back these four lines fixes it for me.

nicolas-grekas replied:
Which one? Something like that: https://github.com/php/php-src/issues/8982?

catch56 replied:
I've put up a very quick PR which resolves the regression for me. We're not using attributes at all yet.

https://github.com/symfony/symfony/pull/47461
